### ISSUE TITLE
Number of times summarizer refreshes snapshot is too high

### ISSUE THREAD

**Issue by vladsud** [microsoft/FluidFramework#8995]
I do not know how urgent / important this issue is, leaving to Paul to triage and make decision if Feb is appropriate.

Marcelo brought to my attention, that empty meeting notes documents continuously summarize, without any user changes. This looks like FFX bug (tracked separately), so this issue is not about that.
But as it summarizes, it does a lot of fetches of not-latest snapshots.
Looking at sample session, I do see these events pointing to issue:
> fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end {"fetchLatest":false}
fluid:telemetry:OdspDriver:LastSummaryProposedHandleMismatch

Unfortunately meeting notes scenario does not send data to our Aria tenant.
But I see that many summaries hit the same issue - these are counts for 1 day across all scenarios that send data:

> fluid:telemetry:Summarizer:Running:SummaryOp	254132
fluid:telemetry:OdspDriver:LastSummaryProposedHandleMismatch	50648
fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end	19414

It's definitely not normal to see that many cases where we need to refresh snapshots. And it's not clear why driver complains that we are referencing wrong parent snapshot. It would be great to analyze why it happens, and if we see sessions where it happens all the time like in meeting notes scenario

Sample meeting note file that demonstrates the issue:
https://www.onenote.com/notes/c3BvOmh0dHBzOi8vbWljcm9zb2Z0LW15LnNoYXJlcG9pbnQtZGYuY29tL3BlcnNvbmFsL21hcmNtYXNfbnRkZXZfbWljcm9zb2Z0X2NvbQ==/b!lql55Wf6D0qBqgrMMcpaiDshzUolNf1DofbCZqkstIk8wHnov-bCQ6pHMry6gSbx/016K3AXJF6AK7FAKU6YZGZ5BGBIDNXWPFW

**Comment by pleath** [comment]
In all the cases of LastSummaryProposedHandleMismatch that I've looked at, the handle comparison fails because OdspSummaryUploadManager.lastSummaryProposalHandle is empty, which seems to indicate that this instance of the upload manager hasn't successfully uploaded a summary before. In all these cases, I see no uploadSummary event prior to the LastSummaryProposedHandleMismatch for the given container. So does this indicate a problem? I'm not sure the event is telling us something useful. There's a comment about caches, but I don't see any difference in behavior as a result of the failed handle comparison.

**Comment by pleath** [comment]
RefreshGetLatestSnapshot does seem to indicate that there are summary acks that are newer than the snapshot that was used to load the summarizer client. I don't know why that would be, though it doesn't seem to be directly related to the summarizer itself. I've seen similar behavior while investigating SummaryMetadataMismatch errors here: https://github.com/microsoft/FluidFramework/issues/8066. The snapshot returned from a getSnapshot call doesn't always seem to be the last one that was ack'd.

**Comment by vladsud** [comment]
@pleath, I'm not sure all the issues here are resolve. 
fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end  shows up in 7% of summaries, and that's quite an additional load on service (and extra latency and bandwidth for client). I think we need to understand it better why it happens and what we can do about it (or conclude why it's hard to improve and that we are Ok with extra cost).

**Comment by pleath** [comment]
Right. I still need to find out why Container's are loaded from older snapshots. The reason I let this issue close is that RefreshLatestSummaryGetSnapshot is also at the heart of the SummaryMetadataMismatch bug. I'll note the linkage between the 2 issues in that one.

**Comment by vladsud** [comment]
So, I think that means this issue should be active, right? Such that we have something to track it?
I do not want to tie it to other bugs that track other kind of telemetry errors, even if it turns out to be same issue.
I'd rather track the core issue here, and if we have enough info, dedup bugs later.

**Comment by agarwal-navin** [comment]
For the `fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end {"fetchLatest":false}` log, we should also add the refSeq# and handles of the downloaded snapshot. That might give some insights into which of a previous summary ack we just got and can help us connect some dots.

**Comment by agarwal-navin** [comment]
Added the above properties - https://github.com/microsoft/FluidFramework/pull/9225

**Comment by marcmasmsft** [comment]
@pleath are there updates on this issue? Anecdotally, I'm seeing this on scriptor files very recently created. For instance, from a file I created today:

<img width="295" alt="image" src="https://user-images.githubusercontent.com/47159987/158266906-45048be2-b569-429f-af86-0f39a85d1bc0.png">

**Comment by marcmasmsft** [comment]
from the last 7 days in SPDF:

<img width="236" alt="image" src="https://user-images.githubusercontent.com/47159987/158271348-99fb65ff-490a-4b02-bb79-d047dde8cc81.png">

For .fluid files, about 10% of the requests to the server were not using the latest keyword
For .loop and .note, that goes up to about 28%
For .whiteboard files, a much healthier 0.4%

**Comment by pleath** [comment]
Interesting data. I haven't looked at this lately, because we're waiting for the logging that Navin recently added to show up in telemetry, which should happen after 0.58 is released.

**Comment by NicholasCouri** [comment]
Documenting what I'm seeing on code + telemetry . 
fetchLatest == true will only happen when the ISummarizeOptions.refreshLatestAck == true which only happens in the case where there is an Heuristics summarize attempt. Then again, it will only happen in the second attempt (unless it gets cancelled):
packages\runtime\container-runtime\src\runningSummarizer.ts@344
 const attempts: (ISummarizeOptions & { delaySeconds?: number })[] = [
                { refreshLatestAck: false, fullTree: false },
                **{ refreshLatestAck: true, fullTree: false },**
                { refreshLatestAck: true, fullTree: false, delaySeconds: 2 * 60 },
                { refreshLatestAck: true, fullTree: true, delaySeconds: 10 * 60 },

Notice summaryAttempts can go up to 8 as we consider retries in the attempts. 
Here is the telemetry data backing it up: 

union Office_Fluid_FluidRuntime_*
| where Data_hostCorrelationId !in (InvalidatedHostCorrelationIds() )
| where Data_eventName contains "RefreshLatestSummaryGetSnapshot"  
| summarize count() by Data_fetchLatest , Data_summaryAttempts, substring(Data_eventName, strlen("fluid:telemetry:Summarizer:"))

<HTML>
		

Data_fetchLatest | Data_summaryAttempts | Column1 | count_
-- | -- | -- | --
0 |   | RefreshLatestSummaryGetSnapshot_cancel | 903
0 |   | Running:RefreshLatestSummaryGetSnapshot_cancel | 62
0 |   | RefreshLatestSummaryGetSnapshot_end | 2093145
0 |   | Running:RefreshLatestSummaryGetSnapshot_end | 630
1 |   | Running:RefreshLatestSummaryGetSnapshot_cancel | 2754
1 | 3 | Running:RefreshLatestSummaryGetSnapshot_cancel | 337
1 | 2 | Running:RefreshLatestSummaryGetSnapshot_cancel | 671
1 | 4 | Running:RefreshLatestSummaryGetSnapshot_cancel | 73
1 | 5 | Running:RefreshLatestSummaryGetSnapshot_cancel | 1
1 | 6 | Running:RefreshLatestSummaryGetSnapshot_cancel | 1
1 |   | Running:RefreshLatestSummaryGetSnapshot_end | 13108
1 | 2 | Running:RefreshLatestSummaryGetSnapshot_end | 3123
1 | 3 | Running:RefreshLatestSummaryGetSnapshot_end | 497
1 | 4 | Running:RefreshLatestSummaryGetSnapshot_end | 234
1 | 6 | Running:RefreshLatestSummaryGetSnapshot_end | 6
1 | 5 | Running:RefreshLatestSummaryGetSnapshot_end | 12
1 | 7 | Running:RefreshLatestSummaryGetSnapshot_end | 4
1 | 8 | Running:RefreshLatestSummaryGetSnapshot_end | 3


		</HTML>

**Comment by NicholasCouri** [comment]
Here are the scenarios for fetchLatest: 

1) fetchLatest: false

While the summarizer is running, handleSummaryAcks goes into a loop while the summarizer is running, and keeps waiting for Summary Ack Ops (updates as well when we send the summary and wait for the ack back)
and when it does get it, it uses the proposal handle  (ack.summaryOp.contents.handle) and the summary ackHandle (ack.summaryAck.contents.handle)   to retrieve the 
Snapshot from the Storage starting from the version id ackHandle: 

getVersions(ackHandle)
        return [{ id: versionId, treeId: FileSnapshotReader.FileStorageVersionTreeId }];

2) fetchLatest == true 

Summarize -> SummarizeCore -> submitSummaryCallback  == async (...args) => this.internalsProvider.submitSummary(...args), 

There are 2 scenarios where we can call refreshLatestSummaryAckFromServer to actually get the latest when ISubmitSummaryOptions.refreshLatestAck == true: 


a) Heuristics is called with trySummarize: Idle, MaxTime, Ops  (notice refreshLatestAck == true only after the first attempt with fetchLatest == false fails) . 
b) TrySummarizeOnce - what Summarizer is doing its last attempt before stopping 


When ackHandle is null (fetchLatest == true) -> we retrieves all versions of the document starting from he head ( if fetchLatest == false, we would get 
all the versions of the document starting from the specified versionId - ackHandle)

Looking at the events and Avg time: 

union Office_Fluid_FluidRuntime_*
| where Data_hostCorrelationId !in (InvalidatedHostCorrelationIds() )
| where Data_eventName contains "RefreshLatestSummaryGetSnapshot_end"  
| summarize round(avg(Data_duration)) , count() by Data_fetchLatest , Data_summaryAttempts, substring(Data_eventName, strlen("fluid:telemetry:Summarizer:"))



<HTML>
		

Data_fetchLatest | Data_summaryAttempts | Column1 | avg_Data_duration | count_
-- | -- | -- | -- | --
0 |   | RefreshLatestSummaryGetSnapshot_end | 764 | 2097646
0 |   | Running:RefreshLatestSummaryGetSnapshot_end | 1595 | 630
1 |   | Running:RefreshLatestSummaryGetSnapshot_end | 14279 | 13119
1 | 3 | Running:RefreshLatestSummaryGetSnapshot_end | 11781 | 499
1 | 2 | Running:RefreshLatestSummaryGetSnapshot_end | 26394 | 3132
1 | 4 | Running:RefreshLatestSummaryGetSnapshot_end | 4512 | 234
1 | 5 | Running:RefreshLatestSummaryGetSnapshot_end | 10200 | 12
1 | 6 | Running:RefreshLatestSummaryGetSnapshot_end | 3292 | 6
1 | 7 | Running:RefreshLatestSummaryGetSnapshot_end | 748 | 4
1 | 8 | Running:RefreshLatestSummaryGetSnapshot_end | 317 | 3


		</HTML>

I believe we the name latestFetch is misleading. Not sure we want to change the current behavior taking in consideration it is much faster. I could try it out (inverting the heuristic attempts data) but I don't think it would make much difference. There is also the fact that we would have an additional call to the server to get the latest summary reference sequence number (WaitingForSeq).

One thing I noticed, and it happens in both scenarios ( refreshLatestSummaryAck and refreshLatestSummaryAckFromServer)  is that we do not have the count of scenarios where I'm the summarizer and I simply skip the refresh  (refreshLatestSummaryFromPending - we start at the end of submit summary).  


Also notice (and I need to look at the previous code) but it seems the sessions in which fetLatest == true and isnull(summaryAttempts) happen in older versions of the runtime: 



union Office_Fluid_FluidRuntime_*
| where Data_hostCorrelationId !in (InvalidatedHostCorrelationIds() )
| where Data_eventName contains "RefreshLatestSummaryGetSnapshot_end"  
| where Data_fetchLatest == 1 and isnull(Data_summaryAttempts) 
| summarize count() by Data_runtimeVersion

<HTML>
		

Data_runtimeVersion | count_
-- | --
0.56.9 | 10032
0.55.4 | 2281
0.54.1 | 775
0.56.8 | 17
0.54.3 | 15


		</HTML>

**Comment by NicholasCouri** [comment]
@vladsud   @agarwal-navin  do you mind looking at the comments I left to see if I got it right ? Notice there 2 events 
fluid:telemetry:Summarizer:**RefreshLatestSummaryGetSnapshot_end**
fluid:telemetry:Summarizer:**Running:RefreshLatestSummaryGetSnapshot_end**

**Comment by agarwal-navin** [comment]
> **1) fetchLatest: false**
> 
> While the summarizer is running, handleSummaryAcks goes into a loop while the summarizer is running, and keeps waiting for Summary Ack Ops (updates as well when we send the summary and wait for the ack back) and when it does get it, it uses the proposal handle (ack.summaryOp.contents.handle) and the summary ackHandle (ack.summaryAck.contents.handle) to retrieve the Snapshot from the Storage starting from the version id ackHandle:

`handleSummaryAcks` does not download the snapshot. It calls `refreshLatestSummary` on container runtime which calls `SummarizerNode::refreshLatestSummary`. `SummarizerNode::refreshLatestSummary` will download the snapshot only under the following conditions:
1. It is not tracking the summary, i.e., the summary op corresponding to this ack was not generated by this client and
2. The last summary seen by this client is older than the summary for which the ack was received.

> **2) fetchLatest == true**
> 
> Summarize -> SummarizeCore -> submitSummaryCallback == async (...args) => this.internalsProvider.submitSummary(...args),
> 
> There are 2 scenarios where we can call refreshLatestSummaryAckFromServer to actually get the latest when ISubmitSummaryOptions.refreshLatestAck == true:
> 
> a) Heuristics is called with trySummarize: Idle, MaxTime, Ops (notice refreshLatestAck == true only after the first attempt with fetchLatest == false fails) . b) TrySummarizeOnce - when Summarizer is doing its last attempt before stopping


I don't think b) here is correct. When trySummarize is called on the last attempt, refreshLatestAck is false:
```
this.trySummarizeOnce(
  // summarizeProps
  { summarizeReason: "lastSummary" },
  // ISummarizeOptions, using defaults: { refreshLatestAck: false, fullTree: false }
  {});
```

**Comment by agarwal-navin** [comment]
> I believe the name latestFetch is misleading - perhaps **fechLatestFromhead** ? Anyways, I'm sure we should change the current behavior. Fact is that we always fetch the head when we failed to retrieve based on the handle we have. So we are already having problems. Notice the data shows the latestFetch calls are much faster.

I am not sure I understand. `fetchLatest` says that download the latest snapshot from the server. We do this in 2 scenarios:
1. If we fail to summarize the first time, we try again by fetching latest snapshot from the server. The idea behind this is that if we failed to summarize, things might be seriously wrong with the current state of the system. So, we download the latest stable state, update our local state based on that and attempt another summary starting from this state.
2. When a client gets an ack for a summary for which it did not send the summary op and the last summary it has seen is older than the acked summary.

**Comment by NicholasCouri** [comment]
> `handleSummaryAcks` does not download the snapshot. It calls `refreshLatestSummary` on container runtime which calls `SummarizerNode::refreshLatestSummary`. `SummarizerNode::refreshLatestSummary` will download the snapshot only under the following conditions:
> 
> 1. It is not tracking the summary, i.e., the summary op corresponding to this ack was not generated by this client and
> 2. The last summary seen by this client is older than the summary for which the ack was received.

Yes, I'm assuming that in these 2 events we will not log the event RefreshLatestSummaryGetSnapshot, will we?

**Comment by NicholasCouri** [comment]
> I am not sure I understand. `fetchLatest` says that download the latest snapshot from the server. We do this in 2 scenarios:
> 
> 1. If we fail to summarize the first time, we try again by fetching latest snapshot from the server. The idea behind this is that if we failed to summarize, things might be seriously wrong with the current state of the system. So, we download the latest stable state, update our local state based on that and attempt another summary starting from this state.
> 2. When a client gets an ack for a summary for which it did not send the summary op and the last summary it has seen is older than the acked summary.

My assumption is that, if we log that event RefreshLatestSummaryGetSnapshot is because we will retrieve from the server  either using null as the ackhandle to get the latest version ( fetchLatest == true)  or the existing ackhandle we already have (fetchLatest == false).

**Comment by NicholasCouri** [comment]
> I don't think b) here is correct. When trySummarize is called on the last attempt, refreshLatestAck is false:

Indeed I missed it :) I will update the original comment :)

**Comment by NicholasCouri** [comment]
> * It is not tracking the summary, i.e., the summary op corresponding to this ack was not generated by this client and
> * The last summary seen by this client is older than the summary for which the ack was received.

That is interesting - I will add this description to the original one :) .  Is it correct to assume that a high number of these refreshes (FetchLatest == false) means the Summarizer is probably moving to different clients in around 7% of the sessions ?

**Comment by vladsud** [comment]
I have not looked deep here, but we rarely get to second phase - it's nonissue:

> union Office_Fluid_FluidRuntime_*
| where Data_eventName ==  "fluid:telemetry:Summarizer:Running:Summarize_generate"
| extend First = Data_summarizeCount == 1
| summarize count()
by First, Data_refreshLatestAck 

First | Data_refreshLatestAck | count_
-- | -- | --
false | false | 1,459,011
true | false | 1,094,510
true | true | 1,281
false | true | 1,635
true |   | 416
false |   | 147

And there are 2M of fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end events (all with fetchLatest = false), which greatly exceeds number of first summaries (less than 1.1M), and is not that far from total number of summaries (2.5M), so sounds like 70% of cases are affected. I though (based on my memory ) this number was closer to 30% in the past.

**Comment by vladsud** [comment]
So I do not hit it when using samples.
And looking at the code, this means refreshLatestSummary() calls getSnapshot(), which does not happen in samples.
For me, it exists on first set of IFs:

```ts
        if (proposalHandle !== undefined) {
            const maybeSummaryNode = this.pendingSummaries.get(proposalHandle);

            if (maybeSummaryNode !== undefined) {
                this.refreshLatestSummaryFromPending(proposalHandle, maybeSummaryNode.referenceSequenceNumber);
                return { latestSummaryUpdated: true, wasSummaryTracked: true };
            }
        }
```

Handle for repeated summarize should never be undefined, so I'd guess this.pendingSummaries does not have a summary we are looking for. No clue why. Might be worth adding more telemetry to make sure we have full picture.

BTW, we do see these hits in our automation (Office Fluid Test DB), so possible we can repro it by simply running local test?
Have not looked deeper

**Comment by vladsud** [comment]
Let's look at some sessions:

> union Office_Fluid_FluidRuntime_*
| where Data_containerId == "31f6894b-6522-4d12-89bb-428dbbd25e85"
| project Event_Sequence, Event_Time, Data_eventName, Data_reason, Data_error

Here we see fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end & fluid:telemetry:OdspDriver:fetchSnapshot_end events in a loop!
We got two 412 errors before it started (means we used wrong parent), so maybe they are failing and we just retrying? Not sure. Eventually client recovers.

Data_containerId == "01c7cd56-46e3-4d9d-920e-c20e590bb6b2" is somewhat similar. Client occasionally fails to do fetch 
Data_containerId == "24c95d87-7449-4d10-9c7a-32b981f5fba0" - similar.

Data_containerId == "2f385fa1-141d-4ea1-aa66-c6f6ea50a058" is slightly different. There were two cases where we would submit summary op, but then halfway through waiting for ack (maybe we received ack, but event follows a bit later?) we fetch snapshot.

@NicholasCouri - please glance at individual sessions, as I feel they will have more clues and possibly that data (along with code inspection) would lead to something.
I'd start with cycles of refreshes - this seems really weird (though I'd check if there are other clients for that same doc running at the same time and submitting summaries - if there are, then obviously question - why this client is not exiting, i.e. why do we have two summarizers?)

Can it be recent change in elected summarizer logic being enabled? We should almost never see two summarizers, if we do - that's an issue, they will keep stepping on each other toes!

**Comment by vladsud** [comment]
First case ("31f6894b-6522-4d12-89bb-428dbbd25e85") did not have any parallel summarizers.

But in second case we indeed have multiple summarizers - it's pretty obvious:
> union Office_Fluid_FluidRuntime_*
// | where Data_containerId == "01c7cd56-46e3-4d9d-920e-c20e590bb6b2"
| where Data_docId == "x2M0eWqDMeoPiNYbiDfD0VGpQRc4OWmMdM2mDsC1LkU%3D"
| where Event_Time >= datetime("2022-03-21T07:54:22.747Z")
| where Event_Time <= datetime("2022-03-21T08:20:58.996Z")
| where Data_clientType != "interactive"
| where Data_eventName == "fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end" or Data_eventName == "fluid:telemetry:OdspDriver:uploadSummary_end"
| project Event_Sequence, Event_Time, Data_eventName, Data_reason, Data_error, Data_clientType, Data_containerId
| order by Event_Time asc

Same issue for 3rd one:
> union Office_Fluid_FluidRuntime_*
// | where Data_containerId == "24c95d87-7449-4d10-9c7a-32b981f5fba0"
| where Data_docId == "Xxv2Hz0hgV3Gg6bAQ%2FlDgcMZyIRTiZoZ27sMIf9MjF0%3D"
| where Event_Time >= datetime("2022-03-24T09:38:14.528Z")
| where Event_Time <= datetime("2022-03-24T09:54:45.9Z")
| where Data_eventName == "fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end" or Data_eventName == "fluid:telemetry:OdspDriver:uploadSummary_end"
| where Data_clientType != "interactive"
| project Event_Sequence, Event_Time, Data_eventName, Data_reason, Data_error, Data_clientType, Data_containerId
| order by Event_Time asc

And last one is the same:
> union Office_Fluid_FluidRuntime_*
//| where Data_containerId == "2f385fa1-141d-4ea1-aa66-c6f6ea50a058"
| where Data_docId == "9aNuIZLtn5jtwBtE3YcpW4Xcm8h%2Bm%2BkIgTof8%2FqJjh8%3D"
| where Event_Time >= datetime("2022-03-18T07:32:14.091Z")
| where Event_Time <= datetime("2022-03-18T07:49:29.371Z")
| where Data_eventName == "fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end" or Data_eventName == "fluid:telemetry:OdspDriver:uploadSummary_end"
| where Data_clientType != "interactive"
| project Event_Sequence, Event_Time, Data_eventName, Data_reason, Data_error, Data_clientType, Data_containerId
| order by Event_Time asc

We need to figure out why we have these cases!

**Comment by vladsud** [comment]
"01c7cd56-46e3-4d9d-920e-c20e590bb6b2" had fluid:telemetry:Summarizer:StoppingSummarizer/"parentShouldNotSummarize" event, but stayed online and kept going with fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end only while other client was summarizing. It should have moved to exit, but it did not!
Exactly same for "24c95d87-7449-4d10-9c7a-32b981f5fba0" & "2f385fa1-141d-4ea1-aa66-c6f6ea50a058"

These are 3 containers from respective queries above.

Something prevents them from closing, or exiting runCore(), as that's what closes container.

**Comment by vladsud** [comment]
Here is how many sessions we see where it took over 60 seconds from the time client realized it needs to go away till the moment it actually went away:

let sessions = union Office_Fluid_FluidRuntime_*
| where Data_eventName == "fluid:telemetry:Summarizer:StoppingSummarizer"
| project Data_containerId, Stop=Event_Time;
let end = union Office_Fluid_FluidRuntime_*
| where Data_clientType != "interactive"
| summarize End=max(Event_Time), countif(Data_eventName == "fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end") by Data_containerId;
sessions | join (end) on Data_containerId
| extend diff=(End-Stop)/1s
| where diff > 60
| summarize count(), sum(countif_)

The number is 9131 containers, responsible for only 4559 of fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end themselves. But we do not know how many such events were raised by other containers running in parallel with these containers.

This does not explain 1M number of refreshes, as the max refreshes client has is 12. So that captures only one small class of these problems :(

**Comment by NicholasCouri** [comment]
> I'd guess this.pendingSummaries does not have a summary we are looking for

From my tests, pendingSummaries was always empty. It would take some time for it to store the handle.

**Comment by agarwal-navin** [comment]
The cache thing also explains why some summarizer clients download latest snapshot in what seems like a loop. A summarizer loads from the cache and there have been many summaries after that. It receives the summary acks for all the summaries by the previous clients, downloads each of them and incrementally updates its reference state.

For example, this is a summarizer client that loads from summary with seq# 94 then does a bunch of summaries:
union Office_Fluid_FluidRuntime_*
| where Data_containerId == "65e8b5c4-1307-405f-898c-f8210f9f3b37"
| where Data_eventName == "fluid:telemetry:Summarizer:Running:Summarize_end" or Data_eventName == "fluid:telemetry:Container:Load_end"
| project Event_Time, Data_eventName, Data_sequenceNumber, Data_summarySequenceNumber, Data_referenceSequenceNumber

![image](https://user-images.githubusercontent.com/45832642/163881993-12edecb0-b7b7-4507-9160-68de8a69162f.png)

The next summarizer also loads from summary with seq# 94 from the cache and then downloads snapshot for acks it receives until it finally shuts down:
union Office_Fluid_FluidRuntime_*
| where Data_containerId == "24c95d87-7449-4d10-9c7a-32b981f5fba0"
| where Data_eventName == "fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end" or Data_eventName == "fluid:telemetry:Container:Load_end"
| project Event_Time, Data_eventName, Data_sequenceNumber, Data_summarySequenceNumber, Data_referenceSequenceNumber, Data_fetchLatest

![image](https://user-images.githubusercontent.com/45832642/163882208-d12a0bd9-9358-4eb9-ad15-60ec2f73456d.png)

What makes this worse is that ODSP only stores the last 3 snapshots. So, when the above summarizer client tries to download some of these snapshots, they will fail and this fails the current ongoing summary.

**Comment by agarwal-navin** [comment]
I think Nicolas is right and the summarizer should not load from a cached snapshot. It should always download the latest snapshot to start with. The reason we see such a high percentage of downloading snapshot which is not latest is because more often that not, summarizer loads from a cached snapshot.

There are couple other things we may want to re-think:
1. The last summary - The last summary by a summarizer should only do 1 (or max 2) attempts before it shuts down. The problem here is that when the summarize is in the above state, it will keep retrying for a long time with delays in between and fail every time. That's why for the above document, we see a new summarizer client spawned but the current one keeps attempting to summarize.
2. Summary attempts - When a summary attempt fails, the second one sets `refrestLatestAck = true`. I am not sure if this is the right approach. Summary attempts can fail for any reason and we shouldn't try to download the latest summary by default. We should change the parameters of attempts based on the stage the previous attempt failed at. For example, if it failed before it attempted to upload the summary, chances are that something went wrong locally. We should probably do another attempt with the same params or set fullTree: true and retry. If the upload fails or we get a nack, that's probably an indication that our summary may not be latest and we should download the latest before proceeding.

**Comment by agarwal-navin** [comment]
Another reason why summaries uploaded by a client that loads from cached snapshot never succeeds is because the incremental summary is relative to a much older summary which the server will reject. So, I think the summarizer should not load from cached snapshot.

**Comment by NicholasCouri** [comment]
> 1. The last summary - The last summary by a summarizer should only do 1 (or max 2) attempts before it shuts down.

Just so I can understand, have  we considered not doing any attempts at all in case there is another summarizer starting ?  What would be the problem if we did not do the last summary at all ?

**Comment by vladsud** [comment]
If I remember correctly, timeout event fires non-stop every 20s while summary is in progress. So, it's not that we increase waits 20 seconds a piece and wait 40s. At least that's my memory - please check with code.

refrestLatestAck = true: (again based on my memory looking at telemetry): that does not happen that often and does not contribute to this problem. I'd rather be data driven - maybe it needs to change, but let's open separate issue and articulate impact such that we can decide when to do it.

I've made a note in #9946 RE number of attempts last summarizer should do - agree, it should be limited.

Re using or not using cached snapshot to load:
There are 3.6M summarizer sessions, and 2M of them has RefreshLatestSummaryGetSnapshot_end event, having 1.01 events per session on average. So, it feels like the ideal solution would be to continue to use cached load but find a way for summarizer to refresh its knowledge only once per session, and cheaply (or make get snapshot by ID as cheap as getlatest in SPO).

Here is another possible angle:
Using cached entry is useful if summarizer is loading right after main container loaded. In such case it's very likely that we refresh cache (while loading main container) and loading from cache will do the right thing - no need to load it again.
The part where cache plays bad trick is when summarizer moves around among already running clients because some other client left. So maybe it's all about using much tighter constraints for summarizer in driver RE expiration of snapshot? Not super ideal (as driver should ideally not care about such things), maybe it should be more explicitly exposed in API (there are other reasons to go that route).

**Comment by marcmasmsft** [comment]
We actually have a work item in SPO to make requesting by id essentially the same as requesting latest (when the id requested is latest); however, my understanding is that this call shouldn't even exist.

For example, in sessions with a single client, I see the summarizer requesting the latest snapshot (by id) before making a new summary. Requesting by keyword 'latest' would perform better but ideally there's no reason why the summary should be requested.

**Comment by vladsud** [comment]
@marcmasmsft, here is (based on stats) how best outcome looks like. There are total 3.6M summarizer sessions (in last 30 days).
1. 400K sessions loaded via getLatest call to SPO (not cache). Either cache is slow, or there is no cache.
   - These sessions do not need to refresh their knowledge of snapshot in order to summarize
2. 600K sessions loaded from cache, and cached snapshot was latest, so they are good to go
3. 2.6M sessions loaded from cache, but cached entry was not latest version. Summarizer needs to get latest snapshot tree to operate
   - We do not have cheaper API to read a tree, only full snapshot, thus driver leverages snapshot fetch by ID to get required info.

I'd propose (as a first step here) to leverage cached snapshot, but only if it's no more than 30 seconds old. That way #3 bucket will shrink substantially, while # 1 will increase at the same rate.

@agarwal-navin, can you please remind me why do we need latest tree in the first place? Blob reuse is done via symbolic name (i.e. using handle + path), I believe we do not use blob IDs for that.

**Comment by marcmasmsft** [comment]
@vladsud the parts that are not clear to me are

(1) Why, in a single client scenario, a summarizer would need to retrieve a snapshot from storage (after the initial boot) if that is the only client that would possibly have summarized? and

(2) Why a summarizer would ever make the request by id rather than using keyword 'latest'? My understanding is that the summarizer only cares about 'latest'.

**Comment by vladsud** [comment]
I do not how many clients are in those sessions - it's hard to calculate that.
But even single client who loaded from cached snapshot needs to retrieve snapshot tree for latest snapshot.
It can probably use getLatest flow, it's more of end-to-end problem.

**Comment by agarwal-navin** [comment]
> @agarwal-navin, can you please remind me why do we need latest tree in the first place? Blob reuse is done via symbolic name (i.e. using handle + path), I believe we do not use blob IDs for that.

Not sure if we need the latest tree but do need a recent tree. Couple of reasons why:
1. For nodes (data stores and DDS) that did not change, we send a handle which is basically the path to that node w.r.t. to a previous summary. If the previous summary is not recent, ODSP will not know how to get the summary tree for these nodes.
2. When uploading a summary to storage, we specify the last ack handle we know about. ODSP will most likely reject it if its no a recent summary that it has.

For both these scenarios, I am not sure if ODSP requires the previous summary to be just recent or the latest that ODSP knows about. @marcmasmsft can you please let me know if this understanding is correct?

**Comment by agarwal-navin** [comment]
> (1) Why, in a single client scenario, a summarizer would need to retrieve a snapshot from storage (after the initial boot) if that is the only client that would possibly have summarized? and

Even in a single client scenario, if the client reloads (tab refresh for example), it can load from cache. This snapshot can be quite old (in terms of the number of summaries), so it needs to download a more recent summary so it can send summaries w.r.t. to this summary.
 
> (2) Why a summarizer would ever make the request by id rather than using keyword 'latest'? My understanding is that the summarizer only cares about 'latest'.

The idea is that if the summarizer receives an ack for a summary that is newer that the latest summary it know about, it will download the summary corresponding to that ack. It then updates its latest summary state based on this summary and subsequent summaries are going to be relative to this summary.
However, it may be possible to always download the latest summary if we see a newer ack. So, if there are multiple acks that come in, it will only download the snapshot for the first one it receives and becomes up to date.

**Comment by agarwal-navin** [comment]
> I'd propose (as a first step here) to leverage cached snapshot, but only if it's no more than 30 seconds old. That way #3 bucket will shrink substantially, while # 1 will increase at the same rate.

To be clear, this is to optimize cases where a document is re-opened after all the clients have left. In this case, the summarizer will spawn with the first client and can re-use the snapshot downloaded by this client?

**Comment by agarwal-navin** [comment]
> > I'd propose (as a first step here) to leverage cached snapshot, but only if it's no more than 30 seconds old. That way #3 bucket will shrink substantially, while # 1 will increase at the same rate.
> 
> @vladsud To be clear, this is to optimize cases where a document is re-opened after all the clients have left. In this case, the summarizer will spawn with the first client and can re-use the snapshot downloaded by this client?

The reason I am asking this is because in this case, 30 seconds may not be enough. It will depend on the initial delay before we spawn the summarizer client. We will probably have to wait until that much time before we stop using the cache.

**Comment by vladsud** [comment]
There are 3 cases:
1. 30 second (or 60 seconds): we continue to use cached entry, nothing changes in behavior
2. We are beyond this timeout, and cached entry is not the latest entry - we will not make system worse (as loading from latest will make number of calls to storage same or less, overall).
3. Case in between - we are past timeout, but no more summaries were posted.
   - That's where we will make it worse

If I got it right, using 60 second threshold, # 1 is ~ 600K sessions over last 30 days. # 2 & # 3 are 1.2M sessions.
There are 2M of fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end events, so overall it feels like it should reduce number of calls to storage. This, assuming that most of fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end calls are due to this problem.

I look at it as a simple first step here that makes system slightly better. I do not know what shell we do over long term.

**Comment by agarwal-navin** [comment]
> If I got it right, using 60 second threshold, # 1 is ~ 600K sessions over last 30 days. # 2 & # 3 are 1.2M sessions. There are 2M of fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end events, so overall it feels like it should reduce number of calls to storage. This, assuming that most of fluid:telemetry:Summarizer:RefreshLatestSummaryGetSnapshot_end calls are due to this problem.

How did you get this data? Can you please share the query?
I am not sure if this completely answers my question though. I am wondering if we should the initial delay before starting the summarizer client into account for determining the cache expiry time. I am not sure what the current delay is. There is a throttler that does some calculations but it seems like the max delay is 30 seconds?
What if this goes beyond the cache expiry? The summarizer will never load from cache in these scenarios right?

**Comment by vladsud** [comment]
Delay is 5 seconds (based on looking at telemetry some time ago), Yes, there is some throttler logic, but I never saw it kicking in.

The analyzes here is very inexact. And I think I was a bit off above.

This will give us subset of sessions that have boot event and start summarizer:

> union Office_Fluid_FluidRuntime_*
| summarize countif(Data_eventName == "fluid:telemetry:SummaryManager:CreatingSummarizer"), countif(Data_eventName contains "Container:Load_end") by Data_containerId
| where countif_ > 0 and countif_1 > 0
| summarize count(), sum(countif_)

913,116 | 2,953,283	 

This will give us all CreatingSummarizer that are within 60 seconds from boot:

> let lastSummary = union Office_Fluid_FluidRuntime_*
| where Data_eventName == "fluid:telemetry:SummaryManager:CreatingSummarizer"
| project start=Event_Time, Data_containerId;
let boot = union Office_Fluid_FluidRuntime_*
| where Data_eventName contains "Container:Load_end"
| where Data_clientType == "interactive"
| project boot=Event_Time, Data_containerId;
lastSummary | join kind=inner (boot) on Data_containerId
| summarize count(), countif((start-boot)/1s < 60)

2,958,263	 | 613,063

Note that out of 913K main container sessions that started summarizer, 613K summarizer sessions were within 60s from main container boot. This includes mostly first summarizer session for given container, but will include some number of second summarizer sessions.

What this number does not include is how far Nth summarizer session boot is from N-1 summarizer session boot (within same session), as they also can be within 60 seconds, and thus reuse snapshot updated by previous session.

I think this query captures that (it's very expensive, I'm not sure there is cheaper way to collect that):

> let lastSummary = union Office_Fluid_FluidRuntime_*
| where Data_eventName == "fluid:telemetry:SummaryManager:CreatingSummarizer"
| project start=Event_Time, Data_containerId, Event_Time;
let boot = union Office_Fluid_FluidRuntime_*
| where Data_eventName == "fluid:telemetry:SummaryManager:CreatingSummarizer"
| where Data_clientType == "interactive"
| project boot=Event_Time, Data_containerId;
lastSummary | join kind=leftouter (boot) on Data_containerId
| where start > boot
| summarize duration=min(start - boot) by Data_containerId, Event_Time
| summarize count(), countif((duration)/1s < 60)

2,468,415	| 111, 213

So roughly 613K + 111K sessions are within 60 seconds from some previous boot. Out of 2.9M sessions we are looking at.


This is where cache is used today. These numbers can't be directly compared to above (only proportions), as we excluded a ton of sessions above that did not have boot event for main container.

> union Office_Fluid_FluidRuntime_*
| where Data_eventName contains "ObtainSnapshot_end"
| where Data_clientType != "interactive"
| summarize count() by Data_method

method | count
-|-
cache | 3382930
network | 552865

We can adjust timeouts and see if we get better hit rate (i.e. use 2 minutes, for example).

**Comment by vladsud** [comment]
Ops, accidently clicked wrong button.
